### PR TITLE
Change the method of calculating the block timestamp

### PR DESCRIPTION
**Pull request by TrustHenry**
Agora changed the method of calculating the block timestamp.

### CODE REVIEW

**Comment by MichaelKim20** [comment]
If you expect a timestamp to be included in the block next time, it seems that there are too many changes.
If the timestamp is not included in the future, I think it's better to decide whether the block is newly created or restored.

File: src/modules/storage/LedgerStorage.ts [hunk 1]
```diff
@@ -583,20 +590,13 @@ export class LedgerStorage extends Storages {
...
         const now_timestamp = Math.floor(new Date().getTime() / 1000);
-
-        // In the production environment, it should be treated as the received time.
-        if (process.env.NODE_ENV === "production") {
-            block.header.time_offset = now_timestamp - genesis_timestamp;
-        } else {
-            block.header.time_offset = JSBI.toNumber(block.header.height.value) * this.block_interval;
-        }
-
+        block.header.time_offset = now_timestamp - this.genesis_timestamp;
```
Comment (hewison-chris) [inline src/modules/storage/LedgerStorage.ts]: What will this time_offset be used for?

Comment (hewison-chris) [reply src/modules/storage/LedgerStorage.ts]: I think it should also be set to the calculated offset based on the block height.

**Comment by MichaelKim20** [comment]
> If you expect a timestamp to be included in the block next time, it seems that there are too many changes. If the timestamp is not included in the future, I think it's better to decide whether the block is newly created or restored.

Of course, this is what Henry and I decided to do on RLT.

**Comment by MichaelKim20** [comment]
Agora will always make the blocks exactly at time intervals, but assuming that exceptional situations can arise, I suggest the following way.

Case 1) When a new block is received:

	block.header.time_offset = Math.floor(new Date().getTime() / 1000) - genesis_timestamp;

Case 2) When a recovery block is received:

        Height of recent block: LH
        TimeStamp of recent block: LT

        Height of recovery block: RH
        TimeStamp of recovery block: RT

        RT = (RH - LH) * Block_Interval + LT
        block.header.time_offset = RT - genesis_timestamp;


경우 1) 신규 블럭이 수신되었을 때 :

	block.header.time_offset = Math.floor(new Date().getTime() / 1000) - genesis_timestamp;
	
경우 2) 복구 블럭이 수신되었을 때 :

	최근 블록의 높이 : LH
	최근 블록의 타임스탬프 : LT
	
	복구 블록의 높이 : RH
	복구 블록의 타임스탬프 : RT

	RT = (RH - LH) * Block_Interval + LT

	block.header.time_offset = RT - genesis_timestamp;

**Comment by TrustHenry** [comment]
@hewison-chris I'd like to hear @hewison-chris' opinion on Michael's comment.

**Comment by hewison-chris** [comment]
I still think that we should use a consistent timestamp for blocks that is only calculated from the block height. If we start creating timestamps based on the clock time on different servers which might receive the block at different times it will lead to confusion. When you think about it the actual time a block is externalized can be different on every node. The transactions that are included were posted at various times during the block interval so each `tx` has in theory a different timestamp, but we have a general timestamp which is based on the intended block interval.

**Comment by MichaelKim20** [comment]
> I still think that we should use a consistent timestamp for blocks that is only calculated from the block height. If we start creating timestamps based on the clock time on different servers which might receive the block at different times it will lead to confusion. When you think about it the actual time a block is externalized can be different on every node. The transactions that are included were posted at various times during the block interval so each `tx` has in theory a different timestamp, but we have a general timestamp which is based on the intended block interval.

Your suggestion makes sense. First, let's revise the code according to your opinion. And let's keep checking in the future.

**Comment by TrustHenry** [comment]
I applied @hewison-chris @MichaelKim20 as a suggestion.

**Review by hewison-chris — APPROVED** [review]
LGTM

**Review by MichaelKim20 — APPROVED** [review]
LGTM
You've done a lot of hidden work.

**Comment by TrustHenry** [comment]
Thank you for the review.

**Pull request merged**